### PR TITLE
feat: support decorator chaining and class creation in ES5

### PR DESCRIPTION
**Pull request by mhevery**
This PR allows to write code in ES5 like so:

``` javascript
Component({
  selector: 'my-component'
}).
View({
  template: 'Hello {{name}}'
}).
Class({
  constructor: [Greeter, Query(Type), function(greeter, types) {
    this.name = 'world';
    this.greeter = greeter;
  }],
  greet: function() {
    this.greeter.greet(this.name);
  }
})
```

### CODE REVIEW

**Comment by mhevery** [comment]
@bradlygreen @davideast @rkirov

File: modules/angular2/src/transform/common/annotation_matcher.dart [hunk 1]
```diff
@@ -15,7 +15,7 @@ const INJECTABLES = const [
   const AnnotationDescriptor(
       'Injectable', 'package:angular2/src/di/annotations_impl.dart', null),
   const AnnotationDescriptor(
-      'Injectable', 'package:angular2/src/di/decorators.dart', null),
+      'Injectable', 'package:angular2/src/di/decorators_spec.dart', null),
```
Comment (vicb) [inline modules/angular2/src/transform/common/annotation_matcher.dart]: ?

File: modules/angular2/src/util/decorators.ts [hunk 2]
```diff
@@ -1,30 +1,64 @@
...
+}
+
+export function Class(clsDef: ClassDefinition): Type {
+  var constructor = clsDef.constructor;
+  var proto = constructor.prototype;
+  if (typeof clsDef.extends == 'function') {
+    constructor.prototype = proto = Object.create((<Function>clsDef.extends).prototype);
+  };
+  for (var key in clsDef) {
+    if (key != 'extends' && clsDef.hasOwnProperty(key)) {
```
Comment (vicb) [inline modules/angular2/src/util/decorators.ts]: I think we should also exclude `prototype`, I had an issue with Safari related to that: https://github.com/angular/zone.js/commit/b91f8fec98591d1ab75f948119130c570baf16af#diff-50e3aa130a4f97a42ee2cf111c7b1d9dR152

File: modules/angular2/test/util/decorators_spec.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+export function main() {
+  var Reflect = global.Reflect;
+
+  var TerminalDecorator = makeDecorator(TerminalAnnotation, null);
+  var TestDecorator = makeDecorator(TestAnnotation, (fn: any) => fn.Terminal = TerminalDecorator);
+  var TestParamDecorator = makeParamDecorator(TestAnnotation);
+
+  describe("decorators", () => {
+    it('invoke as decorator', () => {
```
Comment (vicb) [inline modules/angular2/test/util/decorators_spec.ts]: should

File: modules/angular2/src/transform/common/annotation_matcher.dart (hunk 1 above)
Comment (mhevery) [inline modules/angular2/src/transform/common/annotation_matcher.dart]: Webstorm must have gotten confused on rename. :-) reverted.

File: modules/angular2/src/util/decorators.ts (hunk 2 above)
Comment (mhevery) [inline modules/angular2/src/util/decorators.ts]: Good idea, added check and test.

File: modules/angular2/test/util/decorators_spec.ts (hunk 3 above)
Comment (mhevery) [inline modules/angular2/test/util/decorators_spec.ts]: fixed

**Comment by mhevery** [comment]
@vicb added support for parameter annotations. Could you have a look.

File: modules/angular2/src/core/annotations/decorators.ts [hunk 4]
```diff
@@ -7,14 +12,40 @@ import {
   UnboundedAnnotation
 } from './visibility';
 import {AttributeAnnotation, QueryAnnotation} from './di';
-import {makeDecorator, makeParamDecorator} from '../../util/decorators';
+import {makeDecorator, makeParamDecorator, TypeDecorator} from '../../util/decorators';
+import {Type} from 'angular2/src/facade/lang';
+import {Class} from '../../util/decorators';
```
Comment (vicb) [inline modules/angular2/src/core/annotations/decorators.ts]: merge with `makeDecorator`, ...

File: modules/angular2/src/core/annotations/decorators.ts [hunk 5]
```diff
@@ -7,14 +12,40 @@ import {
...
+
+export interface DirectiveTypeDecorator extends TypeDecorator {}
+
+export interface ComponentTypeDecorator extends TypeDecorator {
+  View(obj: ViewArgs): ViewTypeDecorator;
```
Comment (vicb) [inline modules/angular2/src/core/annotations/decorators.ts]: view ? (lowercase v)

File: modules/angular2/src/util/decorators.ts [hunk 6]
```diff
@@ -1,48 +1,149 @@
...
+import {global, Type, isFunction, stringify} from 'angular2/src/facade/lang';
 
-export function makeDecorator(annotationCls) {
-  return function(...args) {
-    var Reflect = global.Reflect;
-    if (!(Reflect && Reflect.getMetadata)) {
-      throw 'reflect-metadata shim is required when using class decorators';
+export interface ClassDefinition {
+  extends?: Type;
+  constructor: (Function | Array<any>);
```
Comment (vicb) [inline modules/angular2/src/util/decorators.ts]: I'm not sure if ctor are supported in ifs ?

File: modules/angular2/src/util/decorators.ts [hunk 7]
```diff
@@ -1,48 +1,149 @@
...
+  if (fnOrArray === Object || fnOrArray === String || fnOrArray === Function ||
+      fnOrArray === Number || fnOrArray === Array) {
+    throw new Error(`Can not use native ${stringify(fnOrArray)} as constructor`);
+  }
+  if (isFunction(fnOrArray)) {
+    return <Function>fnOrArray;
+  } else if (fnOrArray instanceof Array) {
+    var annotations: Array<any> = fnOrArray;
+    var fn: Function = fnOrArray[fnOrArray.length - 1];
+    if (typeof fn !== 'function') {
```
Comment (vicb) [inline modules/angular2/src/util/decorators.ts]: isFunction (consistency)

File: modules/angular2/test/util/decorators_spec.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+export function main() {
+  var Reflect = global.Reflect;
+
+  var TerminalDecorator = makeDecorator(TerminalAnnotation, null);
+  var TestDecorator = makeDecorator(TestAnnotation, (fn: any) => fn.Terminal = TerminalDecorator);
+  var TestParamDecorator = makeParamDecorator(TestAnnotation);
+
+  describe("decorators", () => {
+    it('shoulld invoke as decorator', () => {
```
Comment (vicb) [inline modules/angular2/test/util/decorators_spec.ts]: should, single quote above

File: modules/angular2/src/core/annotations/decorators.ts [hunk 9]
```diff
@@ -7,14 +12,40 @@ import {
...
+  (obj: ViewArgs): ViewTypeDecorator;
+  new (obj: ViewArgs): ViewAnnotation;
+}
+
 
 /* from annotations */
-export var Component = makeDecorator(ComponentAnnotation);
-export var Directive = makeDecorator(DirectiveAnnotation);
+export var Component = <Component>makeDecorator(ComponentAnnotation, (fn: any) => fn.View = View);
+export var Directive = <Directive>makeDecorator(DirectiveAnnotation, null);
```
Comment (vicb) [inline modules/angular2/src/core/annotations/decorators.ts]: not sur if typing the var would have the same effect as the cast. If it has, I would prefer it.

File: modules/angular2/src/util/decorators.ts [hunk 10]
```diff
@@ -1,48 +1,149 @@
...
-  return function(...args) {
-    var Reflect = global.Reflect;
-    if (!(Reflect && Reflect.getMetadata)) {
-      throw 'reflect-metadata shim is required when using parameter decorators';
+var Reflect = global.Reflect;
+if (!(Reflect && Reflect.getMetadata)) {
+  throw 'reflect-metadata shim is required when using class decorators';
+}
+
+export function makeDecorator(annotationCls, chainFn: (fn: Function) => void): (...args) =>
```
Comment (vicb) [inline modules/angular2/src/util/decorators.ts]: should chainFn be optional, seems like some call sites have not been updated with `null`

File: modules/angular2/src/core/annotations/decorators.ts [hunk 11]
```diff
@@ -7,14 +12,40 @@ import {
...
+
+export interface ComponentTypeDecorator extends TypeDecorator {
+  View(obj: ViewArgs): ViewTypeDecorator;
+}
+
+export interface ViewTypeDecorator extends TypeDecorator { View(obj: ViewArgs): ViewTypeDecorator }
+
+export interface Directive {
+  (obj: any): DirectiveTypeDecorator;
+  new (obj: DirectiveAnnotation): DirectiveAnnotation;
```
Comment (vicb) [inline modules/angular2/src/core/annotations/decorators.ts]: obj: ...Args ? same below

**Comment by vicb** [comment]
The one thing I don't like is how chaining is hard-coded. It might be nicer for competition but it makes harder to add support for other annotations ie RouteConfig

File: modules/angular2/src/core/annotations/decorators.ts (hunk 4 above)
Comment (mhevery) [inline modules/angular2/src/core/annotations/decorators.ts]: I think you are looking at stale SHA, code is no longer there.

File: modules/angular2/src/util/decorators.ts (hunk 6 above)
Comment (mhevery) [inline modules/angular2/src/util/decorators.ts]: Not sure I understand your comments. The code is typed and it passes `tsc` so it seems to be supported.

File: modules/angular2/src/util/decorators.ts (hunk 7 above)
Comment (mhevery) [inline modules/angular2/src/util/decorators.ts]: fixed

File: modules/angular2/test/util/decorators_spec.ts (hunk 8 above)
Comment (mhevery) [inline modules/angular2/test/util/decorators_spec.ts]: fixed

File: modules/angular2/src/core/annotations/decorators.ts (hunk 9 above)
Comment (mhevery) [inline modules/angular2/src/core/annotations/decorators.ts]: `var x:string = 5` will fail so the cast really needs to be no LHS. `var x = <string>5`

File: modules/angular2/src/util/decorators.ts (hunk 10 above)
Comment (mhevery) [inline modules/angular2/src/util/decorators.ts]: Made optional

**Comment by mhevery** [comment]
Not sure I understand hard coded. Any annotations such as `RouterConfig` can supply additional function to add chaining code.

`var RouteConfig = makeDecorator(RouteConfigAnnotation, (fn) {
  fn.Component = Component;
  fn.Custum = CustomFunctionRef;
});

File: modules/angular2/src/util/decorators.ts (hunk 6 above)
Comment (vicb) [inline modules/angular2/src/util/decorators.ts]: forget about it, it's a property, not the ctor method

**Comment by vicb** [comment]
@mhevery I am not really familiar with this code but I was under the impression that `Component({}).View({}).RouteConfig({})` would not be possible unless you modify `View` to add a `RouteConfig` property to the  returned object - and may be add it also to the object returned by `Component` to enable `Component({}).RouteConfig({}).View({})` but then you would have to add `View` to the object returned by `RouteConfig`, ...

File: modules/angular2/test/util/decorators_spec.ts [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+      expect(chain.annotations[0].arg.marker).toEqual('WORKS');
+      expect(chain.annotations[1] instanceof TerminalAnnotation).toEqual(true);
+    });
+
+    describe('Class', () => {
+      it('should create a class', () => {
+        var i0, i1;
+        var MyClass = Class({
+          extends: Class({
+            constructor: function() {},
```
Comment (IgorMinar) [inline modules/angular2/test/util/decorators_spec.ts]: you seem to support inheritance but there doesn't seem to be a way to call the parent constructor. is that intentional?

**Pull request closed without merging**